src/cleanfigure.m

```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+  % will be needed to lookup position of updated points
+  heap = (2:n-1)';
+  len = numel(heap);
+  pos = (1:len)';
+
+
+  area = @(i,j,k) abs((y(j) - y(k)).*x(i) + (y(k)-y(i)).*x(j) + ...
+                                                    (y(i)-y(j)).*x(k))/2;
+
+  % Endpoints are assigned infinte area so they can't be removed
+  a = [Inf;area((1:n-2)', (2:n-1)', (3:n)')';Inf];
+
+  % Keep track of the maximum area removed so far to ensure 
+  % a given element will only be excluded after earlier elements
+  maxArea = 0;
+
+  %Heapify the 'heap' array based on the area, using Floyd's alg
+  root = bitshift(len,-1); %starting with the first parent node
+  while root >= 1
+    down(root);
+    root = root-1;
+  end
+  
+
+  %Now remove the element with the minimum area off the heap
+  while len > 1
+     if a(heap(1)) > maxArea
+       maxArea = a(heap(1));
+     else
+       % ensure current element will only be excluded when earlier elements
+       % are also excluded
+       a(heap(1)) = maxArea;
+     end
+
+    % remove smallest element from heap
+    e = pop(1);
+
+   
+    left = llst(e,1);
+    right = llst(e,2);
+    % remove element from linked list
+    llst(left,2) = llst(e,2);
+    llst(right,1) = llst(e,1);
+
+
+    %Update area of neighbouring points if they're not the ends points
+    if llst(left,1) > 0
+      a(left) = area(llst(left,1),left,llst(left,2));
+      pop(pos(left-1));
+      push();
+    end
+
+    if llst(right,2) > 0
+      a(right) = area(llst(right,1),right,llst(right,2));
+      pop(pos(right-1));
+      push();
+    end
+  end
+
+  if numel(heap) >0 &&  a(heap(1)) < maxArea
+    a(heap(1)) = maxArea;
+  end
+
+  % Move element at "root" down the heap, assuming the heap property
+  % is satisfied for the rest of the tree
+  function down(root)
+    while 2*root  <= len %while the root has a child
+      lchild = 2*root; 
+      rchild = lchild+1;
+      % Find the minimum of the root and its children
+      swap = root;
+      if a(heap(lchild)) < a(heap(swap)) ...
+          || (a(heap(lchild)) == a(heap(swap)) && rand(1)>0.5)
+        swap = lchild;
+      end
+      if rchild <= len && (a(heap(rchild)) < a(heap(swap)) ...
+          || (a(heap(rchild)) == a(heap(swap)) && rand(1)>0.5))
+        swap = rchild;
+      end
+
+      % if the root is the minimum, then we're done
+      if swap == root
+        break
+        % otherwise, swap the root and its minimum child
+      else
+        %swap in the pos array
+        pos(heap(root)-1) = swap;
+        pos(heap(swap)-1) = root;
+
+        %swap in the heap array
+        t = heap(root);
+        heap(root) = heap(swap);
+        heap(swap) = t;
+
+
+        root = swap;
+      end
+    end
+  end
+
+  function up(child)
+    while child > 1
+      parent = bitshift(n,-1);
+      if a(heap(child)) < a(heap(parent)) ...
+              || (a(heap(child)) == a(heap(parent)) && rand(1)>0.5)
```

Comment:
What's that `rand` doing here?
Is it really essential that the algorithm is randomized? If not: please hard-code a value, since such random behavior makes testing much harder.